### PR TITLE
daemon: make socket split backward-compatible.

### PR DESCRIPTION
**Pull request by kyrofa**

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -181,21 +181,38 @@ func (d *Daemon) Init() error {
 		return err
 	}
 
-	if len(listeners) != 2 {
-		return fmt.Errorf("daemon does not handle %d listeners right now, only two", len(listeners))
+	listenersCount := len(listeners)
+	if listenersCount > 2 {
```
Comment (niemeyer) [inline daemon/daemon.go]: As mentioned earlier in the channel, let's please drop this. This breaks forward compatibility (future daemon supports more).

File: daemon/daemon.go [hunk 2]
```diff
@@ -181,21 +181,38 @@ func (d *Daemon) Init() error {
...
+		return fmt.Errorf("daemon can only handle a maximum of 2 listeners right now (got %d)", listenersCount)
 	}
 
-	listenerMap := map[string]net.Listener{
-		listeners[0].Addr().String(): listeners[0],
-		listeners[1].Addr().String(): listeners[1],
+	listenerMap := make(map[string]net.Listener)
```
Comment (niemeyer) [inline daemon/daemon.go]:
```
for _, l := range listeners {
        listenerMap[l.Addr().String()] = l
}

File: daemon/daemon.go [hunk 3]
```diff
@@ -181,21 +181,38 @@ func (d *Daemon) Init() error {
...
+	} else {
+		return fmt.Errorf("daemon is missing the listener for %s", dirs.SnapdSocket)
+	}
 
-	// Note that the SnapSocket listener does not use ucrednet. We use the lack
-	// of remote information as an indication that the request originated with
-	// this socket.
-	d.snapListener = listenerMap[dirs.SnapSocket]
+	// The SnapSocket is not required.
+	if listener, ok := listenerMap[dirs.SnapSocket]; ok {
```
Comment (niemeyer) [inline daemon/daemon.go]: This should do the right thing even if the socket is missing (I think, must confirm due to implicit typing):
```
d.snapListener = listenerMap[dirs.SnapSocket]

File: tests/main/snapctl/task.yaml [hunk 4]
```diff
@@ -73,3 +73,18 @@ execute: |
...
+    /lib/systemd/systemd-activate -E SNAPD_DEBUG=1 -l /run/snapd.socket /usr/lib/snapd/snapd &
+    pid=$!
+
+    export SNAP_CONTEXT="foo"
+    if snap run --hook=apply-config snapctl-hooks; then
+        echo "Expected hook to fail, but it succeeded"
+        exit 1
+    fi
+
+    kill -15 $pid
```
Comment (niemeyer) [inline tests/main/snapctl/task.yaml]: What happens if the test fails early?

File: daemon/daemon.go (hunk 1 above)
Comment (kyrofa) [inline daemon/daemon.go]: Done.

File: daemon/daemon.go (hunk 2 above)
Comment (kyrofa) [inline daemon/daemon.go]: Yeah that was ugly. Not sure what I was thinking. Done.

**Comment by niemeyer** [comment]
LGTM!

**Pull request merged**